torch/csrc/Module.cpp

```diff
@@ -192,30 +192,66 @@ PyObject * THPModule_fromNumpy(PyObject *_unused, PyObject *array)
  * STATELESS FUNCTIONS
  **/
 
+static PyObject * findTensor(PyObject *args, PyObject *kwargs) {
+  for (int i = 0; i < PyTuple_Size(args); i++) {
+    PyObject *item = PyTuple_GET_ITEM(args, i);
+    if (THPModule_isTensor(item) || THPVariable_Check(item)) {
+      return item;
+    }
+  }
+  if (kwargs) {
+    Py_ssize_t pos = 0;
+    PyObject *key, *value;
+    while (PyDict_Next(kwargs, &pos, &key, &value)) {
+      if (THPModule_isTensor(value) || THPVariable_Check(value)) {
+        return value;
+      }
+    }
+  }
+  return THPDefaultTensorClass;
+}
+
+static PyObject * swapFirstTwoItems(PyObject *args) {
+  // Returns a tuple with the first two items swapped
+  auto size = PyTuple_GET_SIZE(args);
+  auto r = THPObjectPtr{PyTuple_New(size)};
+  if (!r) return nullptr;
+  for (Py_ssize_t i = 0; i < size; i++) {
+    PyObject* obj = PyTuple_GET_ITEM(args, (i <= 1 ? 1 - i : i));
+    Py_INCREF(obj);
+    PyTuple_SET_ITEM(r.get(), i, obj);
+  }
+  return r.release();
+}
+
+static PyObject * dispatchStateless(PyObject *args, PyObject *kwargs, const char *name) {
+  PyObject *tensor = findTensor(args, kwargs);
+  return THPUtils_dispatchStateless(tensor, name, args, kwargs);
+}
+
+static PyObject * dispatchStatelessAddXX(PyObject *args, PyObject *kwargs, const char *name) {
+  PyObject *tensor = findTensor(args, kwargs);
+  if (THPVariable_Check(tensor) && PyTuple_GET_SIZE(args) >= 2 && tensor == PyTuple_GET_ITEM(args, 1)) {
+    // On Variables, swap the first two arguments if the 'self' argument comes
+    // second. This handles the deprecated torch.addxx signatures. For example,
+    // torch.addmm(1, var, 2, a, b) -> var.addmm(1, 2, a, b)
+    auto newArgs = THPObjectPtr{swapFirstTwoItems(args)};
```

Comment:
omg what a hack. Why does it need the order swapped? Is it how ATen API looks like or what?

Reply:
The ATen API looks like:

```
Tensor addmm(Tensor self, Tensor mat1, Tensor mat2, Scalar beta=1, Scalar alpha=1);
```

We've made  `beta` and `alpha` keyword-only in Python, which moves them to the end of the C++ signature. (We still support the old-signatures, but they're marked deprecated, see below).

The Python parsing looks like:

```
  static PythonArgParser parser({
    "addmm(Tensor mat1, Tensor mat2, *, Scalar beta=1, Scalar alpha=1)",
    "addmm(Scalar beta, Scalar alpha, Tensor mat1, Tensor mat2)|deprecated",
    "addmm(Scalar beta, Tensor mat1, Tensor mat2)|deprecated",
  });
```

The problem is that we Variable.addmm as always an *instance* method. This means `self`
has to come first, or Python will barf that you're trying to call it on a non-Variable object.

In the Tensor world, we also had a `THPTensor_stateless_(addmm)` which accepted `beta` before `self/source`. We don't have that in the Variable/ATen world. We either use static methods (`cat`) or dispatch to the instance method (`addmm` and most functions).